### DISCUSSION TITLE
Timestepping comparisons

### DISCUSSION THREAD

**Discussion by ftalbrecht** [pymor/pymor#1530] [Ideas]
This is mainly a note to myself to time different step functions for the the new timesteppers, to be introduced in #1529. Since the performance of the steppers does not really matter for big models, I will use small FOMs and ROMs with a large number of time steps to judge the performance of different implementations of the step function.

### results

All tests performed on my notebook (attached to power) within a jupyter notebook, using FOMs and ROMs with 11 DoFs and `2**16 = 65536` time steps.

| scenario | 1: `before` (s) | 2: `dynamic` (s) | 3: `mimic before` (s) | 4: `dynamic + prealloc` (s) |
| ---- |----: | ----: | ----: | ----: |
| explicit FOM | 1.617 | 43.545 | 44.531 | 4.620 |
| explicit ROM | 1.405 | 5.809 | 6.187 | 4.398 |
| implicit FOM | 5.173 | 46.282 | - | 8.679 |
| implicit ROM | 4.872 | 10.097 | - | 8.384 |

### variants of `_step_function`

1. `before`: stepping from [main at e2aad83f](https://github.com/pymor/pymor/blob/e2aad83ff29728ff22be0aad796166504a8de702/src/pymor/algorithms/timestepping.py#L180) before the refactoring
2. `dynamic`: the original approach [in #1529 at 52718274](https://github.com/pymor/pymor/blob/527182742eb40eccb0a07f01bfe23d1ccfde0c77/src/pymor/algorithms/timestepping.py#L432) where the step function was dynamically created on iterator init using complicated branching in init to avoid branching during the run
3. `mimic before`: stepping from 88191047e3e6babc0e589ebc630fb1648d84b968, mimicing the behaviour from 1, including missing treatment of non-trivial mass matrices
4. `dynamic + prealloc`: stepping from 6deed440195f16dcccc6b83521b488c971e6a081 same as `dynamic`, but with the correct pre-allocation of all memory required to store the solution trajectory, as in the old `before` approach

#### explicit FOM

```python
import time
import numpy as np
from pymor.basic import *
from pymor.analyticalproblems.instationary import InstationaryProblem

p = InstationaryProblem(StationaryProblem(
    CircleDomain(),
    advection=ConstantFunction(dim_domain=1, value=np.array([1,]))),
    initial_data=ExpressionFunction(dim_domain=1, expression='1.*(0.1 <= x[0])*(x[0] <= 0.2)'),
    T=1,
    name='linear_transport')
fom, _ = discretize_instationary_fv(p, diameter=1/10, nt=2**16)
fom = fom.with_time_stepper(interpolation='P0')

tic = time.time()
U = fom.solve()
tic = time.time() - tic
print(f'solving FOM with {fom.solution_space.dim} DoFs and {fom.time_stepper.nt} time steps took {tic}s')
```

#### explicit ROM

using the explicit FOM from above
```python
from pymor.algorithms.hapod import inc_hapod

RB, _, _ = inc_hapod(20, U, eps=1e-4, omega=0.1)
reductor = InstationaryRBReductor(fom, RB[:10])
rom = reductor.reduce()

tic = time.time()
u = rom.solve()
tic = time.time() - tic
print(f'solving ROM with {fom.solution_space.dim} DoFs and {rom.time_stepper.nt} time steps took {tic}s')
```

#### implicit FOM

```python
import time
from pymor.basic import *
from pymor.analyticalproblems.instationary import InstationaryProblem

p = InstationaryProblem(StationaryProblem(
    LineDomain(),
    diffusion=ConstantFunction(1, dim_domain=1),
    rhs=ConstantFunction(1, dim_domain=1)),
    initial_data=ConstantFunction(0, dim_domain=1),
    T=10,
    name='heat_equation')
fom, _ = discretize_instationary_cg(p, diameter=1/10, nt=2**16)
fom = fom.with_time_stepper(interpolation='P0')

tic = time.time()
U = fom.solve()
tic = time.time() - tic
print(f'solving FOM with {fom.solution_space.dim} DoFs and {fom.time_stepper.nt} time steps took {tic}s')
```

#### implicit ROM

using the implicit FOM from above
```python
from pymor.algorithms.hapod import inc_hapod

RB, _, _ = inc_hapod(20, U, eps=1e-4, omega=0.1)
reductor = InstationaryRBReductor(fom, RB[:10])
rom = reductor.reduce()

tic = time.time()
u = rom.solve()
tic = time.time() - tic
print(f'solving ROM with {fom.solution_space.dim} DoFs and {rom.time_stepper.nt} time steps took {tic}s')
```

**Comment by renefritze** [comment]
Right align and trim the table would be nice IMO:

|scenario	|1: `before`(s)	|2: `dynamic`(s)|
| ----------- | -----: | ----: |
|explicit FOM	|1.62|	43.54|
|explicit ROM|	1.40	|5.81|
|implicit FOM|	5.17|46.28|
|implicit ROM|	4.87	|10.10|

**Reply by ftalbrecht** [reply]
thx for the hint, done

**Comment by ftalbrecht** [comment]
First take away comparing scenarios 1, 2 and 3. It does make a difference to put the branching into the init, out of the run. However, compared to scenario 1, results are ridiculously extremely disheartening, @sdrave. I'll try to debug more :/ Any recommendations of good profiling of pure Python code?

**Reply by renefritze** [reply]
Builtin cprofile can output stats to be consumed with kcachegrind (which I think you've used before, right?).
Newer (>2019?) versions of Intel's Vtune can directly run python code and I like the UI a lot better than kcachegrind.

**Reply by renefritze** [reply]
Also, for your other/rough/scoped timing needs, may I recommend [this library](https://pytimings.readthedocs.io/en/latest/examples/basic.html) :wink:

**Reply by sdrave** [reply]
There's also lineprofiler which works quite well. I'd be quite surprised if this is really only due to Python overhead. I would expect that more work is done in some way.

**Reply by ftalbrecht** [reply]
Since regular profiling found `numpy`s `ufunc` to take up most of the runtime, the `line-profiler` proved quite helpful, thx!

**Comment by ftalbrecht** [comment]
The issue was a missing `reserve` when initializing the array for the solution trajectory, together with `NumpyVectorArray`s `append`. The performance of variant 4 (the new design + the old pre-allocation) is much better. I will use 4 as a baseline to try some other improvements and the impact of the `_step_function` simplification.

What is you opinion regarding further performance improvements, @sdrave @renefritze? The new design is 2 to 3 times slower than current main (in the corner cases I chose above), while providing more functionality (iterators and interpolation).

**Reply by renefritze** [reply]
Do you have a rough idea how large the problem needs to be in your setup so that the stepper code is less than 1% runtime per step?

**Reply by renefritze** [reply]
Also, how does the stepper overhead scale when keeping time step count the same, but increase DoF count?

**Reply by sdrave** [reply]
A slowdown factor of 3 is quite substantial considering the fact that 11 dim. ROMs are not unheard of. I think it would be good to gain a better understanding of what contributes most to the slowdown. (Maybe by removing some features in a debug branch?)
